larrecodnn/ImagePatternAlgs/NuSonic/Triton/TritonClient.cc

```diff
@@ -0,0 +1,301 @@
+      continue;
+    auto [curr_itr, success] = output_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(oname), std::forward_as_tuple(oname, nicOutput, noBatch_));
+    auto& curr_output = curr_itr->second;
+    outputsTriton_.push_back(curr_output.data());
+    if (verbose_) {
+      io_msg << "  " << oname << " (" << curr_output.dname() << ", " << curr_output.byteSize()
+             << " b) : " << triton_utils::printColl(curr_output.shape()) << "\n";
+    }
+    if (!s_outputs.empty())
+      s_outputs.erase(oname);
+  }
+
+  //check if any requested outputs were not available
+  if (!s_outputs.empty())
+    throw cet::exception("MissingOutput")
+        << "Some requested outputs were not available on the server: " << triton_utils::printColl(s_outputs);
+
+  //propagate batch size to inputs and outputs
+  setBatchSize(1);
+
+  //print model info
+  std::stringstream model_msg;
+  if (verbose_) {
+    model_msg << "Model name: " << options_.model_name_ << "\n"
+              << "Model version: " << options_.model_version_ << "\n"
+              << "Model max batch size: " << (noBatch_ ? 0 : maxBatchSize_) << "\n";
+    MF_LOG_INFO("TritonClient") << model_msg.str() << io_msg.str();
+  }
+}
+
+bool TritonClient::setBatchSize(unsigned bsize) {
+  if (bsize > maxBatchSize_) {
+    MF_LOG_WARNING("TritonClient") << "Requested batch size " << bsize << " exceeds server-specified max batch size "
+                                   << maxBatchSize_ << ". Batch size will remain as" << batchSize_;
+    return false;
+  } else {
+    batchSize_ = bsize;
+    //set for input and output
+    for (auto& element : input_) {
+      element.second.setBatchSize(bsize);
+    }
+    for (auto& element : output_) {
+      element.second.setBatchSize(bsize);
+    }
+    return true;
+  }
+}
+
+void TritonClient::reset() {
+  for (auto& element : input_) {
+    element.second.reset();
+  }
+  for (auto& element : output_) {
+    element.second.reset();
+  }
+}
+
+bool TritonClient::getResults(std::shared_ptr<nic::InferResult> results) {
+  for (auto& [oname, output] : output_) {
+    //set shape here before output becomes const
+    if (output.variableDims()) {
+      std::vector<int64_t> tmp_shape;
+      bool status = triton_utils::warnIfError(results->Shape(oname, &tmp_shape),
+                                              "getResults(): unable to get output shape for " + oname);
+      if (!status)
+        return status;
+      output.setShape(tmp_shape, false);
+    }
+    //extend lifetime
+    output.setResult(results);
+  }
+
+  return true;
+}
+
+//default case for sync and pseudo async
+void TritonClient::evaluate() {
+  //in case there is nothing to process
+  if (batchSize_ == 0) {
+    finish(true);
+    return;
+  }
+
+  // Get the status of the server prior to the request being made.
+  const auto& start_status = getServerSideStatus();
+
+  //blocking call
+  auto t1 = std::chrono::high_resolution_clock::now();

```

Comment:
Suggest using `std::chrono::steady_clock`, which is guaranteed to have monotonically increasing time.

Reply:
done